File: usb-bench-c/src/w_ctrl.c
/*
 * w_ctrl.c — W-ctrl benchmark: USB control-transfer RTT on an Arduino Nano.
 *
 * Sends GET_DESCRIPTOR(Device) to the device N times and records the RTT
 * (nanoseconds) and resource usage for each transfer.
 *
 * Compiles identically for C1 (native-libusb) and C2 (wasi-libusb):
 *   C1: cmake -B build-native && cmake --build build-native
 *   C2: cmake -B build-wasi -DCMAKE_TOOLCHAIN_FILE=toolchain-wasi.cmake \
 *             && cmake --build build-wasi
 *
 * Usage:
 *   w_ctrl [output.csv] [VID:PID] [iterations]
 *   Defaults: bench_w_ctrl.csv  2341:8057  1000
 */

#include <libusb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "csv.h"

/* ── Default parameters ──────────────────────────────────────────────────── */
#define DEFAULT_OUTPUT     "bench_w_ctrl.csv"
#define DEFAULT_VID        0x2341u   /* Arduino Nano 33 BLE */
#define DEFAULT_PID        0x8057u
#define DEFAULT_ITERATIONS 1000u

/* GET_DESCRIPTOR(Device): bmRequestType=0x80, bRequest=0x06, wValue=0x0100 */
#define GET_DESC_REQ_TYPE  0x80u
#define GET_DESC_REQUEST   0x06u
#define GET_DESC_VALUE     0x0100u
#define GET_DESC_INDEX     0u
#define DEVICE_DESC_LEN    18

/* ── Helpers ─────────────────────────────────────────────────────────────── */

static void die(const char *msg)
{
    fprintf(stderr, "fatal: %s\n", msg);
    exit(1);
}

static void parse_vid_pid(const char *s, uint16_t *vid, uint16_t *pid)
{
    unsigned v = 0, p = 0;
    if (sscanf(s, "%x:%x", &v, &p) != 2)
        die("expected VID:PID in hex, e.g. 2341:8057");
    *vid = (uint16_t)v;
    *pid = (uint16_t)p;
}

/* ── Main ─────────────────────────────────────────────────────────────────── */

int main(int argc, char **argv)
{
    const char *output     = (argc > 1) ? argv[1] : DEFAULT_OUTPUT;
    const char *vid_pid    = (argc > 2) ? argv[2] : "2341:8057";
    uint32_t    iterations = (argc > 3) ? (uint32_t)atoi(argv[3]) : DEFAULT_ITERATIONS;

    uint16_t vid, pid;
    parse_vid_pid(vid_pid, &vid, &pid);

    /* ── libusb init ─────────────────────────────────────────────────────── */
    libusb_context *ctx = NULL;
    int rc = libusb_init(&ctx);
    if (rc != 0)
        die("libusb_init failed");

    libusb_device_handle *handle = libusb_open_device_with_vid_pid(ctx, vid, pid);
    if (!handle) {
        fprintf(stderr, "fatal: device %04x:%04x not found\n", vid, pid);
        libusb_exit(ctx);
        return 1;
    }

    /* ── CSV logger ─────────────────────────────────────────────────────── */
    FILE *csv = bench_csv_open(output);
    if (!csv)
        die("bench_csv_open failed");

    fprintf(stderr,
            "▶ %04x:%04x  workload=ctrl  iterations=%u  "
            "condition=" BENCH_CONDITION "  → %s\n",
            vid, pid, iterations, output);

    /* ── Benchmark loop ──────────────────────────────────────────────────── */
    uint8_t desc_buf[DEVICE_DESC_LEN];

    for (uint32_t iter = 0; iter < iterations; ++iter) {
        bench_rusage_t ru_before, ru_after;
        bench_rusage_now(&ru_before);

        /* ── timed section ─────────────────────────────────────────────── */
        uint64_t t0 = bench_now_ns();

        int n = libusb_control_transfer(
            handle,
            GET_DESC_REQ_TYPE,
            GET_DESC_REQUEST,
            GET_DESC_VALUE,
            GET_DESC_INDEX,
            desc_buf,
            (uint16_t)DEVICE_DESC_LEN,
            500 /* timeout ms */
        );

        uint64_t duration_ns = bench_now_ns() - t0;
        /* ── end timed section ─────────────────────────────────────────── */

        bench_rusage_now(&ru_after);

        if (n < 0) {
            fprintf(stderr,
                    "  [%4u/%u] libusb_control_transfer error: %s\n",
                    iter, iterations,
                    libusb_error_name(n));
            continue;
        }

        bench_row_t row = {0};
        row.condition    = BENCH_CONDITION;
        row.workload     = "ctrl";
        row.iteration    = iter;
        row.bytes        = (uint64_t)n;
        row.duration_ns  = duration_ns;
        row.user_cpu_us  = ru_after.user_us  - ru_before.user_us;
        row.sys_cpu_us   = ru_after.sys_us   - ru_before.sys_us;
        row.rss_peak_kb  = ru_after.rss_peak_kb;
        row.guest_mem_bytes = bench_guest_mem_bytes();
        row.has_guest_mem   = 0; /* native: emit empty column */
#ifdef __wasm__
        row.has_guest_mem   = 1;
#endif
        bench_csv_write(csv, &row);

        if (iter % 100 == 0 || iter == iterations - 1) {
            fprintf(stderr,
                    "  [%4u/%u]  RTT=%llu µs  bytes=%d\n",
                    iter, iterations,
                    (unsigned long long)(duration_ns / 1000u),
                    n);
        }
    }

    bench_csv_close(csv);
    libusb_close(handle);
    libusb_exit(ctx);

    fprintf(stderr,
            "✓ Done  iterations=%u  → %s\n",
            iterations, output);
    return 0;
}

File: usb-bench-rs/src/bin/w_ctrl.rs
//! W-ctrl benchmark — USB control-transfer RTT on an Arduino Nano.
//!
//! Sends GET_DESCRIPTOR(Device) repeatedly and records the per-transfer RTT.
//!
//! # Conditions served by this binary
//! | Build target        | Condition | Default `--condition` |
//! |---------------------|-----------|-----------------------|
//! | native host         | C3        | `native-rusb`         |
//! | wasm32-wasip2       | C5        | `wasi-raw-wit`        |
//!
//! # Usage
//! ```text
//! w_ctrl [output.csv] [VID:PID] [iterations] [--condition <name>]
//!
//! Defaults:
//!   output.csv       bench_w_ctrl.csv
//!   VID:PID          2341:8057  (Arduino Nano 33 BLE)
//!   iterations       1000
//!   condition        native-rusb  (native) | wasi-raw-wit  (wasm)
//! ```

use std::path::Path;
use std::time::Instant;

use anyhow::{anyhow, Result};

use usb_bench_rs::csv::{self, Logger, Row, Rusage};

// ---------------------------------------------------------------------------
// Transport selection — the ONLY cfg-gated section in this file
// ---------------------------------------------------------------------------

// C3 (native) and C4 (wasm32-wasip2 + c4-rusb-wasi) → rusb-based native impl
#[cfg(any(not(target_family = "wasm"), feature = "c4-rusb-wasi"))]
use native::CtrlDevice as ActiveDevice;

// C5 (wasm32-wasip2, raw WIT) → WIT-based wasm impl
#[cfg(all(target_family = "wasm", not(feature = "c4-rusb-wasi")))]
use wasm::CtrlDevice as ActiveDevice;

// ---------------------------------------------------------------------------
// Native transport (rusb)
// ---------------------------------------------------------------------------

// C3 (native) and C4 (wasm32-wasip2 + c4-rusb-wasi feature) — rusb impl
#[cfg(any(not(target_family = "wasm"), feature = "c4-rusb-wasi"))]
mod native {
    use anyhow::{anyhow, Result};
    use rusb::{GlobalContext, UsbContext};
    use std::time::Duration;

    pub struct CtrlDevice {
        handle: rusb::DeviceHandle<GlobalContext>,
    }

    impl CtrlDevice {
        pub fn open(vid: u16, pid: u16) -> Result<Self> {
            let ctx = GlobalContext::default();
            let handle = ctx
                .open_device_with_vid_pid(vid, pid)
                .ok_or_else(|| anyhow!("device {:04x}:{:04x} not found", vid, pid))?;
            Ok(CtrlDevice { handle })
        }

        /// GET_DESCRIPTOR(Device) — returns the raw descriptor bytes.
        pub fn get_descriptor(&mut self) -> Result<Vec<u8>> {
            let mut buf = [0u8; 18]; // Device Descriptor is 18 bytes
            let n = self.handle.read_control(
                0x80,   // bmRequestType: device-to-host, standard, device
                0x06,   // bRequest: GET_DESCRIPTOR
                0x0100, // wValue: Device Descriptor type 0x01, index 0
                0,      // wIndex
                &mut buf,
                Duration::from_millis(500),
            )?;
            Ok(buf[..n].to_vec())
        }
    }
}

// ---------------------------------------------------------------------------
// WASM transport (raw WIT bindings) — C5 only
// ---------------------------------------------------------------------------

#[cfg(all(target_family = "wasm", not(feature = "c4-rusb-wasi")))]
mod wasm {
    use anyhow::{anyhow, Result};

    // WIT types from the central bindings module (one generate! per binary).
    use usb_bench_rs::bindings::component::usb::device::{list_devices, DeviceHandle};
    use usb_bench_rs::bindings::component::usb::transfers::{
        await_transfer, TransferOptions, TransferSetup, TransferType,
    };

    pub struct CtrlDevice {
        handle: DeviceHandle,
    }

    impl CtrlDevice {
        pub fn open(vid: u16, pid: u16) -> Result<Self> {
            let devices =
                list_devices().map_err(|e| anyhow!("list_devices: {:?}", e))?;
            let (dev, _, _) = devices
                .into_iter()
                .find(|(_, d, _)| d.vendor_id == vid && d.product_id == pid)
                .ok_or_else(|| anyhow!("device {:04x}:{:04x} not found", vid, pid))?;
            let handle = dev.open().map_err(|e| anyhow!("open: {:?}", e))?;
            Ok(CtrlDevice { handle })
        }

        pub fn get_descriptor(&mut self) -> Result<Vec<u8>> {
            let setup = TransferSetup {
                bm_request_type: 0x80,
                b_request: 0x06,
                w_value: 0x0100,
                w_index: 0,
            };
            let opts = TransferOptions {
                endpoint: 0,
                timeout_ms: 500,
                stream_id: 0,
                iso_packets: 0,
            };
            let xfer = self
                .handle
                .new_transfer(TransferType::Control, setup, 18, opts)
                .map_err(|e| anyhow!("new_transfer: {:?}", e))?;
            xfer.submit_transfer(&[])
                .map_err(|e| anyhow!("submit_transfer: {:?}", e))?;
            let result =
                await_transfer(&xfer).map_err(|e| anyhow!("await_transfer: {:?}", e))?;
            Ok(result.data)
        }
    }
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

fn main() -> Result<()> {
    let cfg = Config::from_args(std::env::args().collect())?;
    let mut logger = Logger::open(Path::new(&cfg.output))?;
    let device = ActiveDevice::open(cfg.vid, cfg.pid)?;
    run_benchmark(device, &cfg, &mut logger)
}

// ---------------------------------------------------------------------------
// Shared benchmark loop (identical for native and wasm)
// ---------------------------------------------------------------------------

fn run_benchmark(
    mut device: ActiveDevice,
    cfg: &Config,
    logger: &mut Logger,
) -> Result<()> {
    eprintln!(
        "▶ {vid:04x}:{pid:04x}  workload=ctrl  iterations={it}  \
         condition={cond}  → {out}",
        vid = cfg.vid,
        pid = cfg.pid,
        it = cfg.iterations,
        cond = cfg.condition,
        out = cfg.output,
    );

    let loop_start = Instant::now();

    for iter in 0..cfg.iterations {
        let ru_before = Rusage::snapshot();

        // ── timed section ────────────────────────────────────────────────────
        let t0 = Instant::now();
        let data = device.get_descriptor()?;
        let duration_ns = t0.elapsed().as_nanos() as u64;
        // ── end timed section ────────────────────────────────────────────────

        let ru_after = Rusage::snapshot();
        let delta = ru_after.delta(&ru_before);

        let mut row = Row::new(cfg.condition.as_str(), "ctrl");
        row.iteration = iter;
        row.bytes = data.len() as u64;
        row.duration_ns = duration_ns;
        row.user_cpu_us = delta.user_us;
        row.sys_cpu_us = delta.sys_us;
        row.rss_peak_kb = delta.rss_peak_kb;
        row.guest_mem_bytes = csv::guest_mem_bytes();
        logger.write(&row)?;

        if iter % 100 == 0 || iter == cfg.iterations - 1 {
            let rtt_us = duration_ns / 1_000;
            eprintln!(
                "  [{iter:4}/{it}]  RTT={rtt_us} µs  bytes={}",
                data.len(),
                it = cfg.iterations,
            );
        }
    }

    let wall_secs = loop_start.elapsed().as_secs_f64();
    eprintln!(
        "✓ Done  iterations={}  wall={wall:.2} s  → {out}",
        cfg.iterations,
        wall = wall_secs,
        out = cfg.output,
    );
    Ok(())
}

// ---------------------------------------------------------------------------
// CLI configuration
// ---------------------------------------------------------------------------

struct Config {
    output: String,
    vid: u16,
    pid: u16,
    iterations: u32,
    condition: String,
}

impl Config {
    fn from_args(args: Vec<String>) -> Result<Self> {
        let output = args
            .get(1)
            .map(String::as_str)
            .unwrap_or("bench_w_ctrl.csv")
            .to_owned();
        let vid_pid = args.get(2).map(String::as_str).unwrap_or("2341:8057");
        let iterations: u32 = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(1000);
        let condition = args
            .windows(2)
            .find(|w| w[0] == "--condition")
            .map(|w| w[1].clone())
            .unwrap_or_else(|| default_condition().to_owned());
        let (vid, pid) = parse_vid_pid(vid_pid)?;
        Ok(Self { output, vid, pid, iterations, condition })
    }
}

fn default_condition() -> &'static str {
    #[cfg(not(target_family = "wasm"))]
    { "native-rusb" }
    #[cfg(all(target_family = "wasm", feature = "c4-rusb-wasi"))]
    { "wasi-rusb" }
    #[cfg(all(target_family = "wasm", not(feature = "c4-rusb-wasi")))]
    { "wasi-raw-wit" }
}

fn parse_vid_pid(s: &str) -> Result<(u16, u16)> {
    let (v, p) = s
        .split_once(':')
        .ok_or_else(|| anyhow!("expected VID:PID (hex), got {s:?}"))?;
    let vid = u16::from_str_radix(v.trim(), 16)
        .map_err(|e| anyhow!("invalid VID {v:?}: {e}"))?;
    let pid = u16::from_str_radix(p.trim(), 16)
        .map_err(|e| anyhow!("invalid PID {p:?}: {e}"))?;
    Ok((vid, pid))
}
